src/main/java/org/openjdk/jextract/clang/Index.java

```diff
@@ -98,6 +99,36 @@ public TranslationUnit parseTU(String file, Consumer<Diagnostic> dh, int options
         }
     }
 
+    public TranslationUnit parseTU(String filename, String content, Consumer<Diagnostic> dh, int options, String... args)
+            throws ParsingFailedException {
+        try (Arena arena = Arena.ofConfined()) {
+            MemorySegment filenameSeg = arena.allocateFrom(filename);
+            MemorySegment contentSeg = arena.allocateFrom(content);
+            MemorySegment cargs = args.length == 0 ? null : arena.allocate(C_POINTER, args.length);
+            for (int i = 0 ; i < args.length ; i++) {
+                cargs.set(C_POINTER, i * C_POINTER.byteSize(), arena.allocateFrom(args[i]));
+            }
+
+            MemorySegment unsavedFile = CXUnsavedFile.allocate(arena);
+            CXUnsavedFile.Filename(unsavedFile, filenameSeg);
+            CXUnsavedFile.Contents(unsavedFile, contentSeg);
+            CXUnsavedFile.Length(unsavedFile, content.length());
+
+            MemorySegment tu = Index_h.clang_createTranslationUnitFromSourceFile(
+                    ptr,
+                    filenameSeg,
+                    args.length,
+                    cargs == null ? MemorySegment.NULL : cargs,
+                    1,
+                    unsavedFile);
+            TranslationUnit rv = new TranslationUnit(tu);
+            // even if we failed to parse, we might still have diagnostics
+            rv.processDiagnostics(dh);
+
+            return rv;
+        }
+    }
```

Comment:
Can't we use `clang_parseTranslationUnit2`? It also supports unsaved files, but also supports returning error codes. I suppose `clang_createTranslationUnitFromSourceFile` return `NULL` if parsing fails.

Reply:
I suggest extending the existing `parseTU` to also accept a list of `Index.UnsavedFile`